src/expression_pruner.cpp

```diff
@@ -0,0 +1,77 @@
+#include "expression_pruner.h"
+#include "ir/visitor.h"
+#include "pruner_util.h"
+
+namespace P4PRUNER {
+
+IR::Node *ExpressionPruner::preorder(IR::Operation_Binary *s) {
+
+    IR::Node *to_remove =
+        rand() < 0.5 ? (IR::Node *)s->left : (IR::Node *)s->right;
```

Comment:
I do not think this will work. The rand() call generates an integer that is larger than 0.5 as far as I remember. Also you need to return a node. 
In addition, instead of casting I would set the function to `const`. So in summary I think it should look like this:
```
const IR::Node *ExpressionPruner::preorder(IR::Operation_Binary *s) {
    auto decision = (double)rand() / RAND_MAX;
    if (decision < 0.33) {
        // return the left-hand side of the expression
        return s->left;
    } else if (decision < 0.66) {
        // return the right-hand side of the expression
        return s->right;
    }
    // do nothing, just return the node
    return s;
}
```
The same applies to the other methods.

Reply:
Another thing I realized we unfortunately have to be much more subtle about this. 
There are binary expressions where this won't work. We have to decide at a case by case basis.

Reply:
So instead of Operation_Binary make this IR::Add or IR::Sub